### DISCUSSION TITLE
Getting started OTA Sever

### DISCUSSION THREAD

**Discussion by markicos** [zigpy/zigpy#721] [Q&A]
Hello,

I need some support for starting using OTA, I am already managing several sensor devices with Zigpy but now I need to start updating those sensor.

Could anyone hold my hand on starting working on this?, a sample code or something make that me understand where to put the code to upgrade, or if it this possible a reference to any documentation any kind of help will be welcomed.

Thank in advance,

**Comment by Adminiuga** [comment]
Check zigpy configuration schema. It has pointers to what it should contain to start it's and location of the OTA folder where it looks for the image files

**Reply by markicos** [reply]
You mean that I should  set the folder when creating new app, adding location where I want to put my code, doesn't it?

something like   `"ota":{ "otau_directory": "/img_floder"},` on the creation of the app will work, 
```
app = await ControllerApplication.new(
        config=ControllerApplication.SCHEMA({
            "database_path": "zigbee.db",
            "device": {
                "path": f'{portcom}',
            },
            "ota":{ "otau_directory": "/img_floder"},
        }),
        auto_form=True,
        start_radio=True,
    )

```
But how do i do to check if a new update should be perform ?

Thanks in advance,

**Reply by puddly** [reply]
Devices usually check in and ask for images, which will cause zigpy to initiate an OTA update if a compatible image exists in that folder. If a device does not ask, you can contact the device to notify it of a new image via the `ota` cluster's `image_notify` command:

```Python
# See section 11.13.3.1 of the ZCL spec for the payload format
await device.endpoints[123].ota.image_notify(0x00, 100)
```

Images known to zigpy can be accessed through the `app` object's `ota` property:

```Python
for key, cached_image in app.ota._image_cache.items():
    # Cached images wrap an existing Image object
    image = cached_image.image
    print(image.header)
```

OTA works but isn't very friendly or configurable at the moment. Suggestions (and pull requests!) are appreciated.

**Reply by markicos** [reply]
Thanks a lot both, now I have something to start, I will keep you updated, again thank you

**Comment by Hedda** [comment]
FYI pipiche38 has begun an interesting project that is trying to get OTA firmware URL via info get from Wireshark Zigbee sniffing:

https://github.com/pipiche38/Capture-OTA-from-Wireshark

https://github.com/pipiche38/Domoticz-Zigate-Wiki/blob/master/en-eng/Corner_Retreiving-Legrand-Firmware.md

**Comment by markicos** [comment]
Hello thank you all,

I have some questions, 


Is there any listener I can add to the app in order to know a device is asking for possibility of update?


I have  a .bin and also .hex file, is there any script already implemented to create the header specified in the ZigBee Cluster Library Specification?

**Reply by puddly** [reply]
1. A `cluster_command` event is fired on the device's `Ota` cluster with `hdr.command_id == 0x01  (query_next_image)`.
2. Zigbee SDKs provide all the tools to generate `.ota` files from what I've seen. But if you're trying to make one by hand, you can construct it:

    ```Python
    from zigpy.ota.image import OTAImage, OTAImageHeader, SubElement, ElementTagId
    
    header = OTAImageHeader(
        upgrade_file_id=OTAImageHeader.MAGIC_VALUE,
        header_version=0x0001,
        header_length=0xFFFF,
        field_control=0b000,
        manufacturer_id=0x1234,
        image_type=0x5678,
        file_version=0xAABB,
        stack_version=0xCCDD,
        header_string=b"some text",
        image_size=0xFFFF,
    )
    header.header_length = len(header.serialize())
    
    image = OTAImage(
        header=header,
        subelements=[
            SubElement(tag_id=ElementTagId.UPGRADE_IMAGE, data=b"your firmware"),
        ]
    )
    
    # TODO: remove the `image.serialize()` length assertion so this works:
    #     image.header.image_size = len(image.serialize())
    # Alternatively, make an `image.recompute_sizes()` method that does this.
    image.header.image_size = image.header.header_length + sum(len(s.serialize()) for s in image.subelements)
    
    assert image.serialize() == OTAImage.deserialize(image.serialize())[0].serialize()
    ```

**Reply by markicos** [reply]
I found in the sdk the way to form a img properly but thanks a lot for your code

**Comment by markicos** [comment]
hello,

I don't know if I am doing something wrong but after create the app like this
```
     app = await ControllerApplication.new(
        config=ControllerApplication.SCHEMA({
            "database_path": "zigbee.db",
            "device": {
                "path": f'{portcom}',
            },
            "ota":{ "otau_directory": ".\img_folder"},
        }),
        auto_form=True,
        start_radio=True,
    )
```

See that 'ota' is in the configuration no images are found in `app.ota._image_cache` despite being one .zigbee image properly formed, is there anything i am missing?

Thanks

**Reply by puddly** [reply]
Have you enabled debug logging for zigpy?

```python
import logging
logging.basicConfig(level=logging.DEBUG)  # or whatever other logging config you use

import zigpy
...
```

It will tell you why the image isn't considered valid, or if it's not being seen due to path issues.

**Reply by markicos** [reply]
Yes I have the logging enabed
I saw this line, and a quiestion came to my mind, is hw version  required?
```
DEBUG:zigpy.ota.provider:ImageKey(manufacturer_id=123, image_type=321): .\img_folder\007B-0141-01030101-good_image.zigbee, version: 16974081, hw_ver: (None, None), OTA string: 
good_image
```
But  when i try to get it like this or
```
for key, cached_image in app.ota._image_cache.items():
        # Cached images wrap an existing Image object
        image = cached_image.image
        print(image.header)
```
or like 
```
        a=await app.ota.get_ota_image(0x123, 0x321)
        print(a.header)
```
 as there is nothing in the dict it is printing nothing 

![image](https://user-images.githubusercontent.com/44501063/117252391-2d8b9c00-ae46-11eb-979e-12b8557b64ec.png)

**Comment by markicos** [comment]
Hello again,

I can't see any clue of why not finding images on the `_image_cache`, as https://github.com/zigpy/zigpy/discussions/721#discussioncomment-702384, 

I tried of adding also an img downloaded from https://github.com/zigpy/zigpy/discussions/660, but same result the img is not found in the `_image_cache`.

Anyone can bring some light on it? 

Thank you,

**Reply by markicos** [reply]
"I was blind but now I see" hahaha, well I saw several problem, the script didn't find the OTA cluster as the .db I was using was no updated, also for finding the _img_cache filled it is mandatory to use the 'image_notify' before you can see it, 

`await device.endpoints[123].ota.image_notify(0x00, 100)`
